### DISCUSSION TITLE
Slack "slash commands" not working with my deployment

### DISCUSSION THREAD

**Discussion by enyo** [vercel/community#1717] [Help] [answered]
### Summary

I configured a slash command with slack, and defined the "Request URL”. When I use any slash command inside slack, a request is sent to the correct URL (I can see it in the logs) but it returns `403`.

<img width="842" alt="Screenshot 2023-03-07 at 17 07 56" src="https://user-images.githubusercontent.com/133277/223479724-21da980e-8ae7-44d0-a40c-90d8a95fdc83.png">

The request actually never reaches my server. I tried adding debug logs but they are never outputted.

I assume that there is some sort of blacklist that blocks the webhook?

I can just create a POST request to the same URL and it works fine.

**Comment by amyegan** [comment]
Hi Matias! Common reasons for a `403: Forbidden` status, are a missing or invalid [access token](https://vercel.com/guides/how-do-i-use-a-vercel-api-access-token) or [deploy hook unique id](https://vercel.com/docs/concepts/deployments/deploy-hooks#security). If you're using any [Advanced Deployment Protection Features](https://vercel.com/docs/concepts/deployments/advanced-deployment-protection#), you may need to configure [Protection Bypass for Automation](https://vercel.com/docs/concepts/deployments/advanced-deployment-protection#protection-bypass-for-automation).

If you need more specific recommendations, please tell me more about how you have configured this slash command and what request it is supposed to trigger. That will help me narrow down potential sources of trouble

**Reply by enyo** [reply]
Thanks for your answer @amyegan 
I have a SvelteKit app deployed on vercel, and I don’t use any deployment protection features. The URL I use is publicly accessible as you can see here: https://www.pausly.app/_webhooks/slack/command-link (only POST is allowed).

```bash
% curl -X POST https://www.pausly.app/_webhooks/slack/command-link

{"ok":true}
```

So I’m not sure what triggers 403 when the request comes from slack.

**Reply by enyo** [reply]
@amyegan is there anything you could help me with here? I’m a bit stuck. Thanks!

**Reply by amyegan** [reply]
Sorry for the delay. I had to take some time off at the end of last week. 🙇 

Do you have CORS headers set? This sounds like it could be preflight 403 CORS error. If you don't already have headers set to allow requests from other domains, you can [add CORS headers](https://vercel.com/guides/how-to-enable-cors) in a single serverless function or in a project configuration file

**Reply by enyo** [reply]
@amyegan no worries. I hope you enjoyed your days off.

Correct me if I’m wrong, but wouldn't the CORS error be on the client (browser) side? Also: this is a Slack webhook request, so I would assume CORS would not be involved here... Is there some way for me to get more data from the logs?

**Reply by enyo** [reply]
My problem is, that this is a black box for me. I see the request (see screenshot of original post) in the vercel logs. But if I simply invoke a POST request myself, I get my log output. But with the request coming in from Slack, it simply says `403` and I can’t debug anything...

Just to make sure I now added `('Access-Control-Allow-Origin', '*’)` to my response, but it didn’t change anything.

**Reply by ESchouten** [reply]
@enyo Have you found the solution? Encountering the same, although a manual send via Postman does work. Also on SvelteKit

**EDIT:**
Found the solution, add Content-Type: application/json header

**Reply by smaeda-ks** [reply]
For SvelteKit, in particular, there's a common pitfall that many people forget about, which is to disable CSRF check:
https://kit.svelte.dev/docs/configuration#csrf

**Reply by enyo** [reply]
Hey @ESchouten yes! As @smaeda-ks pointed out, this is due to the SCRF check that SvelteKit itself does.

**Comment by enyo** [comment, marked as answer]
The fix for this issue is to disable SvelteKit’s CSRF check for the specific route. To avoid disabling this for all routes, this is what I did:

```ts
// src/hooks.server.ts

import { csrf } from '$lib/server/hooks/csrf'
import { sequence } from '@sveltejs/kit/hooks'

export const handle = sequence(
  // This is necessary because otherwise Slack webhooks don't pass.
  csrf({ bypassedPaths: ['/_webhooks/'] })
  // You can add any hook handlers that you might need here
)
```

```ts
// src/lib/server/hooks/csrf.ts

import { error, json, text, type Handle } from '@sveltejs/kit'

/**
 * CSRF protection copied from sveltekit but with the ability to bypass csrf
 * protection for specific routes.
 */
export const csrf =
  ({ bypassedPaths }: { bypassedPaths: string[] }): Handle =>
  async ({ event, resolve }) => {
    const forbidden =
      event.request.method === 'POST' &&
      event.request.headers.get('origin') !== event.url.origin &&
      isFormContentType(event.request) &&
      !bypassedPaths.find((path) => event.url.pathname.startsWith(path))

    if (forbidden) {
      const csrfError = error(
        403,
        `Cross-site ${event.request.method} form submissions are forbidden`,
      )
      if (event.request.headers.get('accept') === 'application/json') {
        return json(csrfError.body, { status: csrfError.status })
      }
      return text(csrfError.body.message, { status: csrfError.status })
    }

    return resolve(event)
  }
const isContentType = (request: Request, ...types: string[]) =>
  types.includes(
    request.headers.get('content-type')?.split(';', 1)[0]?.trim() ?? '',
  )

const isFormContentType = (request: Request) =>
  isContentType(
    request,
    'application/x-www-form-urlencoded',
    'multipart/form-data',
  )
```

**Reply by ESchouten** [reply]
Don't do all that, if your request provides the header `Content-Type: application/json`, CSRF is automatically bypassed.
See [https://kit.svelte.dev/docs/configuration#csrf](https://kit.svelte.dev/docs/configuration#csrf)

**Reply by enyo** [reply]
But is there a way to make Slack’s webhooks send the data in `application/json`?

**Reply by ESchouten** [reply]
Docs suggest this happens https://api.slack.com/messaging/webhooks

**Reply by smaeda-ks** [reply]
That's for "incoming" webhooks where Slack is the receiver of payload and you send it with `application/json`. Here, we're talking about the opposite flow.

https://api.slack.com/interactivity/slash-commands#app_command_handling

> This data will be sent with a `content-type` header set as `application/x-www-form-urlencoded`.
